Mathlib/Data/Set/Pairwise.lean

```diff
@@ -0,0 +1,494 @@
+/-
+Copyright (c) 2017 Johannes Hölzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johannes Hölzl
+
+! This file was ported from Lean 3 source module data.set.pairwise
+! leanprover-community/mathlib commit 207cfac9fcd06138865b5d04f7091e46d9320432
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Logic.Relation
+import Mathlib.Logic.Pairwise
+import Mathlib.Data.Set.Lattice
+
+/-!
+# Relations holding pairwise
+
+This file develops pairwise relations and defines pairwise disjoint indexed sets.
+
+We also prove many basic facts about `Pairwise`. It is possible that an intermediate file,
+with more imports than `Logic.Pairwise` but not importing `Data.Set.Lattice` would be appropriate
+to hold many of these basic facts.
+
+## Main declarations
+
+* `Set.PairwiseDisjoint`: `s.PairwiseDisjoint f` states that images under `f` of distinct elements
+  of `s` are either equal or `Disjoint`.
+
+## Notes
+
+The spelling `s.PairwiseDisjoint id` is preferred over `s.Pairwise Disjoint` to permit dot notation
+on `Set.PairwiseDisjoint`, even though the latter unfolds to something nicer.
+-/
+
+
+open Set Function
+
+variable {α β γ ι ι' : Type _} {r p q : α → α → Prop}
+
+section Pairwise
+
+variable {f g : ι → α} {s t u : Set α} {a b : α}
+
+theorem pairwise_on_bool (hr : Symmetric r) {a b : α} :
+    Pairwise (r on fun c => cond c a b) ↔ r a b := by simpa [Pairwise, Function.onFun] using @hr a b
+#align pairwise_on_bool pairwise_on_bool
+
+theorem pairwise_disjoint_on_bool [SemilatticeInf α] [OrderBot α] {a b : α} :
+    Pairwise (Disjoint on fun c => cond c a b) ↔ Disjoint a b :=
+  pairwise_on_bool Disjoint.symm
+#align pairwise_disjoint_on_bool pairwise_disjoint_on_bool
+
+theorem Symmetric.pairwise_on [LinearOrder ι] (hr : Symmetric r) (f : ι → α) :
+    Pairwise (r on f) ↔ ∀ ⦃m n⦄, m < n → r (f m) (f n) :=
+  ⟨fun h _m _n hmn => h hmn.ne, fun h _m _n hmn => hmn.lt_or_lt.elim (@h _ _) fun h' => hr (h h')⟩
+#align symmetric.pairwise_on Symmetric.pairwise_on
+
+theorem pairwise_disjoint_on [SemilatticeInf α] [OrderBot α] [LinearOrder ι] (f : ι → α) :
+    Pairwise (Disjoint on f) ↔ ∀ ⦃m n⦄, m < n → Disjoint (f m) (f n) :=
+  Symmetric.pairwise_on Disjoint.symm f
+#align pairwise_disjoint_on pairwise_disjoint_on
+
+theorem pairwise_disjoint.mono [SemilatticeInf α] [OrderBot α] (hs : Pairwise (Disjoint on f))
+    (h : g ≤ f) : Pairwise (Disjoint on g) :=
+  hs.mono fun i j hij => Disjoint.mono (h i) (h j) hij
+#align pairwise_disjoint.mono pairwise_disjoint.mono
+
+alias Function.injective_iff_pairwise_ne ↔ Function.injective.pairwise_ne _
+
```

Comment:
```suggestion
```
I figured out why it crashed: this line is already present in `Logic.Pairwise` and I guess Lean 3 somehow didn't care.

Reply:
Ah that's a good find, that should make it way easier to MWE!